### PR TITLE
Add diagnostic for unhandled exception from compiler

### LINKED ISSUE #31720: Catch Any throwables at the PhaseRunner
This is required to stop breaking tooling when a BAD-SAD error is generated from the compiler.

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #31720

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by nadeeshaan** [comment]
@pcnfernando Do we have an associated tooling issue as well? We need to have a fix from the LS as well

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java [hunk 1]
```diff
@@ -27,6 +27,9 @@
  */
 public enum DiagnosticErrorCode implements DiagnosticCode {
 
+    // The member represents a bad sad error
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java:30]: Shall we add a comment here explaining that this is a special diagnostic code.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java [hunk 2]
```diff
@@ -409,7 +413,12 @@ static void generateCodeInternal(ModuleContext moduleContext,
         if (bootstrapLangLibName != null) {
             compilerPhaseRunner.performLangLibBirGenPhases(moduleContext.bLangPackage);
         } else {
-            compilerPhaseRunner.performBirGenPhases(moduleContext.bLangPackage);
+            try {
+                compilerPhaseRunner.performBirGenPhases(moduleContext.bLangPackage);
+            } catch (Throwable t) {
+                compilerPhaseRunner.addDiagnosticForUnhandledException(moduleContext.bLangPackage, t);
+                return;
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:420]: Don't we need to add this error to the bad-sad log?

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java (hunk 1 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java]: Updated in 1bfe3d6700a474472e3d57098ca04df8126d908d

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java (hunk 2 above)
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:420]: Added in 1bfe3d6700a474472e3d57098ca04df8126d908d

**Comment by pcnfernando** [comment]
> @pcnfernando Do we have an associated tooling issue as well? We need to have a fix from the LS as well

Created an issue to track it in https://github.com/ballerina-platform/ballerina-lang/issues/32040

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java [hunk 3]
```diff
@@ -395,7 +395,11 @@ static void compileInternal(ModuleContext moduleContext, CompilerContext compile
         if (bootstrapLangLibName != null) {
             compilerPhaseRunner.performLangLibTypeCheckPhases(pkgNode);
         } else {
-            compilerPhaseRunner.performTypeCheckPhases(pkgNode);
+            try {
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/ModuleContext.java:398]: This try-catch block needs to cover symbol enter phase too (line:392). For instance, https://github.com/ballerina-platform/ballerina-lang/issues/32044 is not covered with this.

**Review by rdhananjaya — APPROVED** [review]

**Comment by rdhananjaya** [comment]
Changes look good to me. Since there is no straightforward way to add test cases to check compiler crashes will merge this without tests. Manually tested with known crashes.

**Pull request merged**